selfie-camera/src/main/java/org/odk/collect/selfiecamera/CaptureSelfieActivity.kt

```diff
@@ -0,0 +1,166 @@
+import androidx.camera.video.VideoRecordEvent
+import androidx.camera.view.PreviewView
+import androidx.core.content.ContextCompat
+import org.odk.collect.androidshared.ui.ToastUtils.showLongToast
+import org.odk.collect.externalapp.ExternalAppUtils
+import org.odk.collect.permissions.PermissionsChecker
+import org.odk.collect.strings.localization.LocalizedActivity
+import java.io.File
+
+class CaptureSelfieActivity : LocalizedActivity() {
+
+    private var recording: Recording? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        if (
+            !PermissionsChecker(this).isPermissionGranted(
+                Manifest.permission.CAMERA,
+                Manifest.permission.RECORD_AUDIO
+            )
+        ) {
+            finish()
+            return
+        }
+
+        requestWindowFeature(Window.FEATURE_NO_TITLE)
+        window.setFlags(
+            WindowManager.LayoutParams.FLAG_FULLSCREEN,
+            WindowManager.LayoutParams.FLAG_FULLSCREEN
+        )
+        setContentView(R.layout.activity_capture_selfie)
+
+        val cameraProvider = ProcessCameraProvider.getInstance(this)
+        cameraProvider.addListener(
+            {
+                if (intent.getBooleanExtra(EXTRA_VIDEO, false)) {
+                    setupVideo(cameraProvider.get())
+                } else {
+                    setupStillImage(cameraProvider.get())
+                }
+            },
+            ContextCompat.getMainExecutor(this)
+        )
+
+        showLongToast(this, R.string.take_picture_instruction)
+    }
+
+    @SuppressLint("MissingPermission") // Checked on Activity launch
+    private fun setupVideo(cameraProvider: ProcessCameraProvider) {
+        val previewView = findViewById<PreviewView>(R.id.preview)
+
+        val preview = Preview.Builder().build()
+        preview.setSurfaceProvider(previewView.surfaceProvider)
+
+        val cameraSelector = CameraSelector.Builder()
+            .requireLensFacing(CameraSelector.LENS_FACING_FRONT)
+            .build()
+
+        val recorder = Recorder.Builder()
+            .setExecutor(ContextCompat.getMainExecutor(this))
+            .build()
+
+        val videoCapture = VideoCapture.withOutput(recorder)
+
+        val outputFile = File(intent.getStringExtra(EXTRA_TMP_PATH), "tmp.mp4")
+        val outputFileOptions = FileOutputOptions.Builder(outputFile).build()
+        previewView.setOnClickListener {
+            recording.let {
+                if (it == null) {
+                    recording = videoCapture.output
+                        .prepareRecording(this, outputFileOptions)
+                        .withAudioEnabled()
+                        .start(ContextCompat.getMainExecutor(this)) { event ->
+                            if (event is VideoRecordEvent.Finalize) {
+                                ExternalAppUtils.returnSingleValue(this, outputFile.absolutePath)
```

Comment:
It initially confused me that we were using a different approach to return data to `FormEntryActivity`. I imagine that this was done to mirror the way data is passed back from the standard camera when taking a photo/video. It felt more important to me to have this consistent with other "external data" Activity objects than with the camera interface ([principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)).